Source/WebCore/page/LocalDOMWindow.cpp

```diff
@@ -2461,6 +2461,12 @@ void LocalDOMWindow::setLocation(LocalDOMWindow& activeWindow, const URL& comple
     if (completedURL.protocolIsJavaScript() && frameElement() && !frameElement()->document().contentSecurityPolicy()->allowJavaScriptURLs(aboutBlankURL().string(), { }, completedURL.string(), frameElement()))
         return;
 
+    // If the loader for activeWindow's frame (browsing context) has no outgoing referrer, set its outgoing referrer
+    // to the URL of its main frame's Document — which is the incumbent global object's associated Document.
+    // https://html.spec.whatwg.org/#location-object-navigate
+    if (activeWindow.frame()->loader().outgoingReferrer().isEmpty())
+        activeWindow.frame()->loader().setOutgoingReferrer(dynamicDowncast<LocalFrame>(page()->mainFrame())->document()->urlForBindings());
```

Comment:
`activeFrame->checkedLoader().setOutgoingReferrer`

Reply:
Also, you can get the main frame like so: `frame->mainFrame()` so you don't have to null check the `page()` too.